added time_as_float() and relevant tests

VcelistMC opened the pull request:
Hi,
I really appreciate you leaving opportunities for beginners like me to contribute. I have made the function as requested and added relevant tests.

VcelistMC committed:
added time_as_float and relevant tests

fhightower commented:
Thank you for the contribution @VcelistMC! I'll look over the code and get it merged!

If I may provide one point of feedback about how to best use Github: typically if you are going to work on a pull-request (PR) that addresses an issue (as in this case where you wrote a PR to address issue #7), it is helpful to say something on the issue first and ask to have the issue assigned to you so that only one person works on a solution at a time. Does that make sense?

Thanks again for contributing 🎉  🚀  !

fhightower commented:
Fixes #7

fhightower commented on democritus_dates/dates_and_times.py:
```diff
@@ -459,3 +459,18 @@ def time_waste(n=3):
+
+def time_as_float(time_string: str) -> float:
+    """converts a given HH:MM time string to float"""
+    try:
+        hours, mintues = list(map(int, time_string.split(":"))) #parse given time string
+    except ValueError:
+        message = f"Invalid time string, ensure that the argument is in HH:MM format. Provided value: {time_string}"
+        raise ValueError(message)
+
+    if hours > 23 or mintues > 59:
+        message = f"Invalid time string, should be between 00:00 and 23:59. Provided value: {time_string}"
+        raise ValueError(message)
```
I like the fact that you included error handling and validation. I also like how each error message shows the `time_string` argument that was passed into the function. This makes it really easy to catch simple errors. Nice work!

fhightower commented on democritus_dates/dates_and_times.py:
```diff
@@ -459,3 +459,18 @@ def time_waste(n=3):
+def time_as_float(time_string: str) -> float:
+    """converts a given HH:MM time string to float"""
+    try:
+        hours, mintues = list(map(int, time_string.split(":"))) #parse given time string
```
I suggest fixing a typo in the `minutes` variable name

```suggestion
        hours, minutes = list(map(int, time_string.split(":"))) #parse given time string
```

fhightower commented on democritus_dates/dates_and_times.py:
```diff
@@ -459,3 +459,18 @@ def time_waste(n=3):
+        message = f"Invalid time string, ensure that the argument is in HH:MM format. Provided value: {time_string}"
+        raise ValueError(message)
+
+    if hours > 23 or mintues > 59:
```
I suggest fixing a typo in the `minutes` variable name

```suggestion
    if hours > 23 or minutes > 59:
```

fhightower commented on democritus_dates/dates_and_times.py:
```diff
@@ -459,3 +459,18 @@ def time_waste(n=3):
+        message = f"Invalid time string, should be between 00:00 and 23:59. Provided value: {time_string}"
+        raise ValueError(message)
+
+    return hours + (mintues/60)
```
I suggest fixing a typo in the `minutes` variable name

```suggestion
    return hours + (minutes/60)
```

fhightower commented on democritus_dates/dates_and_times.py:
```diff
@@ -459,3 +459,18 @@ def time_waste(n=3):
+    """converts a given HH:MM time string to float"""
+    try:
+        hours, mintues = list(map(int, time_string.split(":"))) #parse given time string
+    except ValueError:
+        message = f"Invalid time string, ensure that the argument is in HH:MM format. Provided value: {time_string}"
+        raise ValueError(message)
```
I like how you define the error message in a seperate variable. Some people don't prefer this, but I like it because it makes it easier to change how the error message is used. Nice job!

fhightower replied:
I suggest adding the suggestions I've made below. Adding `from e` to the end of the line where you raise a `ValueError` will make the traceback more clear if anyone does run into this exception.

*Before* adding `from e` the exception will look like:

```
{Traceback removed for brevity}
ValueError: invalid literal for int() with base 10: 'a'

During handling of the above exception, another exception occurred:

{Traceback removed for brevity}
ValueError: foo
```

*After* adding `from e` to the end of the `ValueError` call as suggested below will make the errors look like:

```
{Traceback removed for brevity}
ValueError: invalid literal for int() with base 10: 'a'

The above exception was the direct cause of the following exception:

{Traceback removed for brevity}
ValueError: foo
```

The relationship between the two ValueErrors is made more clear in the second case.

You can read more about the `raise ... from` construction [here](https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement).

```suggestion
    except ValueError as e:
        message = f"Invalid time string, ensure that the argument is in HH:MM format. Provided value: {time_string}"
        raise ValueError(message) from e
```

fhightower reviewed (commented):
I've made some suggestions and comments. Look them over and please let me know if you have any questions!

VcelistMC committed:
Merge remote-tracking branch 'upstream/main' into add-time_as_float

VcelistMC committed:
fixed typos, imporved error clarity

VcelistMC commented:
> it is helpful to say something on the issue first and ask to have the issue assigned to you so that only one person works on a solution at a time. Does that make sense?

Yeah, it makes sense. Your issue was the very first issue I've ever looked at on GitHub and it didn't cross my mind to ask you first (sorry). it was only after looking at more issues on other repos and saw how other people make contributions, I realized that maybe I should've said something first.

Anyways, I've fixed the typos and changed the error handling as you said. Let me know if there's anything else to do.

fhightower commented:
> > it is helpful to say something on the issue first and ask to have the issue assigned to you so that only one person works on a solution at a time. Does that make sense?
> 
> Yeah, it makes sense. Your issue was the very first issue I've ever looked at on GitHub and it didn't cross my mind to ask you first (sorry). it was only after looking at more issues on other repos and saw how other people make contributions, I realized that maybe I should've said something first.
> 
> Anyways, I've fixed the typos and changed the error handling as you said. Let me know if there's anything else to do.

No worries! Happy to have you contributing.

Thanks for the updates and for this PR 🚀  ! I'm merging it in now. I'll have some more issues created for the Democritus project soon, so keep an eye out for those if you are interested.
